cpp/tests/reductions/reduction_tests.cu

```diff
@@ -89,54 +114,274 @@ TYPED_TEST_CASE(ReductionTest, Types);
+    cudf::test::column_wrapper<T> col_nulls = construct_null_column(v, host_bools);
+    gdf_size_type valid_count = col_nulls.size() - col_nulls.null_count();
+    auto replaced_array = replace_nulls(int_values, host_bools, int{0});
+
+    double expected_value_nulls = calc_mean(replaced_array, valid_count);
+    this->reduction_test(col_nulls, expected_value_nulls, true,
+        GDF_REDUCTION_MEAN, GDF_FLOAT64);
+}
+
+
+TYPED_TEST(MultiStepReductionTest, var_std)
+{
+    using T = TypeParam;
+    std::vector<int> int_values({-3, 2,  1, 0, 5, -3, -2, 28});
+    std::vector<bool> host_bools({1, 1, 0, 1, 1, 1, 0, 1});
+
+    auto calc_var = [](std::vector<int>& v, gdf_size_type valid_count){
+        double mean = std::accumulate(v.begin(), v.end(), double{0});
+        mean /= valid_count ;
+
+        double sum_of_sq = std::accumulate(v.begin(), v.end(), double{0},
+            [](double acc, TypeParam i) { return acc + i * i; });
+
+        int ddof = 1;
+        gdf_size_type div = valid_count - ddof;
+
+        double var = sum_of_sq / div - ((mean * mean) * valid_count) /div;
+        return var;
+    };
+
+    // test without nulls
+    std::vector<T> v = convert_values<T>(int_values);
+    cudf::test::column_wrapper<T> col(v);
+
+    double var = calc_var(int_values, int_values.size());
+    double std = std::sqrt(var);
+
+    this->reduction_test(col, var, true, GDF_REDUCTION_VAR, GDF_FLOAT64);
+    this->reduction_test(col, std, true, GDF_REDUCTION_STD, GDF_FLOAT64);
+
+    // test with nulls
+    cudf::test::column_wrapper<T> col_nulls = construct_null_column(v, host_bools);
+    gdf_size_type valid_count = col_nulls.size() - col_nulls.null_count();
+    auto replaced_array = replace_nulls(int_values, host_bools, int{0});
+
+    double var_nulls = calc_var(replaced_array, valid_count);
+    double std_nulls = std::sqrt(var_nulls);
+
+    this->reduction_test(col_nulls, var_nulls, true, GDF_REDUCTION_VAR, GDF_FLOAT64);
+    this->reduction_test(col_nulls, std_nulls, true, GDF_REDUCTION_STD, GDF_FLOAT64);
 }
 
+// ----------------------------------------------------------------------------
+
+template <typename T>
+struct ReductionMultiStepErrorCheck : public ReductionTest<T>
+{
+    ReductionMultiStepErrorCheck(){}
+    ~ReductionMultiStepErrorCheck(){}
+
+    void reduction_error_check(cudf::test::column_wrapper<T> &col,
+        bool succeeded_condition,
+        gdf_reduction_op op, gdf_dtype output_dtype)
+    {
+        const gdf_column * underlying_column = col.get();
+        auto statement = [&]() {
+            cudf::reduction(underlying_column, op, output_dtype);
+        };
+
+        if( succeeded_condition ){
+            CUDF_EXPECT_NO_THROW(statement());
+        }else{
+            EXPECT_ANY_THROW(statement());
+        }
+    }
+};
+
+using AllTypes = testing::Types<
+    int8_t,int16_t, int32_t, int64_t, float, double,
+    cudf::bool8, cudf::date32, cudf::date64, cudf::timestamp, cudf::category>;
```

Comment:
What happens for `cudf::nvstring_category` type columns? (`GDF_STRING_CATEGORY`)

Reply:
Good point, I've added  `cudf::nvstring_category` for AllTypes.
The tests worked well.